lib/kms/wrapper.js

```diff
@@ -355,26 +357,49 @@ class KMS {
     }
 
     static checkHealth(log, cb) {
+        logger.debug('current KMS cache state', { lastChecked: cache.getLastChecked(), result: cache.getResult() });
         if (!client.healthcheck) {
             return cb(null, {
                 [implName]: { code: 200, message: 'OK' },
             });
         }
-        return client.healthcheck(log, err => {
-            const respBody = {};
-            if (err) {
-                respBody[implName] = {
-                    error: err.description,
-                    code: err.code,
-                };
-            } else {
-                respBody[implName] = {
-                    code: 200,
-                    message: 'OK',
-                };
-            }
-            return cb(null, respBody);
-        });
+
+        const shouldRefreshCache = cache.shouldRefresh();
+
+        // healthcheck only if cache needs refreshing
+        if (shouldRefreshCache) {
+            logger.debug('health check for KMS backend');
+            return client.healthcheck(log, (err) => {
+                const respBody = {};
+                if (err) {
+                    // Returning HTTP status code 200 allows CloudServer to start up and listen on port 8000
+                    // when KMS is unavailable, and ensures that the S3 frontend deep healthcheck
+                    // still allow requests to this CloudServer instance. This way, an unavailable KMIP
+                    // host or a connection issue with KMIP will not prevent the normal operation of CloudServer.
+                    respBody[implName] = {
+                        code: 200,
+                        message: 'KMS health check failed',
+                        description: err.description,
+                    };
+                    logger.warn('KMS health check failed', { errorCode: err.code, error: err.description });
+                } else {
+                    respBody[implName] = {
+                        code: 200,
+                        message: 'OK',
+                    };
+                    logger.info('KMS health check succeeded', { respBody });
+                }
+                // Only update the cache on success
+                cache.setResult(respBody);
```

Comment:
It means if there is a failure, kms down or other problem, it's going to retry for every deepHealthcheck (every 2 seconds or 30 seconds) indefinitely.

And if the endpoint is slow there could potentially be multiple health check concurrently.

Maybe we could prevent this from happening, maybe it could be desynchronized from the incoming request, by having kms health check on interval to manage failure retry and the request retrieved only the cached value ?